dataset/variable_reduction.cpp

```diff
@@ -28,23 +37,28 @@ Variable flatten(const VariableConstView &var, const Dim dim,
 
 Variable sum(const VariableConstView &var, const Dim dim,
              const MasksConstView &masks) {
-  if (const auto mask_union = irreducible_mask(masks, dim))
-    return sum(var * ~mask_union, dim);
+  if (const auto mask_union = irreducible_mask(masks, dim)) {
+    const auto maskedVar = applyMask(var, mask_union);
+    return sum(maskedVar, dim);
+  }
```

Comment:
```suggestion
  if (const auto mask_union = irreducible_mask(masks, dim))
    return sum(applyMask(var, mask_union), dim);
```
and similar below, no need for introducing named variable?

Reply:
Done for `sum`, I've left `mean` as is though, since we already introduce another variable and readability starts taking a hit